cpp/open3d/pipelines/registration/Registration.cpp

```diff
@@ -194,152 +192,160 @@ RegistrationResult RegistrationRANSACBasedOnCorrespondence(
-                        break;
-                    }
-                }
-                if (!check) continue;
-                transformation = estimation.ComputeTransformation(
-                        source, target, ransac_corres);
-                check = true;
-                for (const auto &checker : checkers) {
-                    if (checker.get().require_pointcloud_alignment_ &&
-                        !checker.get().Check(source, target, ransac_corres,
+                    if (!checker.get().Check(source, target, ransac_corres,
                                              transformation)) {
                         check = false;
                         break;
                     }
                 }
                 if (!check) continue;
+
                 geometry::PointCloud pcd = source;
                 pcd.Transform(transformation);
-                auto this_result = GetRegistrationResultAndCorrespondences(
-                        pcd, target, kdtree, max_correspondence_distance,
+                auto result = EvaluateRANSACBasedOnCorrespondence(
+                        pcd, target, corres, max_correspondence_distance,
                         transformation);
-                if (this_result.fitness_ > result_private.fitness_ ||
-                    (this_result.fitness_ == result_private.fitness_ &&
-                     this_result.inlier_rmse_ < result_private.inlier_rmse_)) {
-                    result_private = this_result;
-                }
-#pragma omp critical
-                {
-                    total_validation = total_validation + 1;
-                    if (total_validation >= criteria.max_validation_)
-                        finished_validation = true;
+
+                if (result.IsBetterRANSACThan(best_result_local)) {
+                    best_result_local = result;
+
+                    // Update exit condition if necessary
+                    double exit_itr_d =
+                            std::log(1.0 - criteria.confidence_) /
+                            std::log(1.0 - std::pow(result.fitness_, ransac_n));
+                    exit_itr_local =
+                            exit_itr_d < double(criteria.max_iteration_)
+                                    ? static_cast<int>(std::ceil(exit_itr_d))
+                                    : exit_itr_local;
                 }
-            }  // end of if statement
-        }      // end of for-loop
+            }  // if < exit_itr_local
+        }      // for loop
 #pragma omp critical
         {
-            if (result_private.fitness_ > result.fitness_ ||
-                (result_private.fitness_ == result.fitness_ &&
-                 result_private.inlier_rmse_ < result.inlier_rmse_)) {
-                result = result_private;
+            if (best_result_local.IsBetterRANSACThan(best_result)) {
+                best_result = best_result_local;
+            }
+            if (exit_itr_local > exit_itr) {
```

Comment:
I also really don't get this. Why is the shared variable only updated if `exit_itr_local` is larger? Furthermore, is the logging below correct, in view of my comment above regarding the use of different stopping criteria among different threads?

Reply:
In general, in this implementation, each thread runs its own allocated iterations (a portion of preset 10K iterations). The results are only synchronized at the end of their loop. So it is correct that each thread is using different criteria, and the final exit condition is a rough estimation. The best result, however, is the best we have.

We can use more complex blocking strategies to make results perfectly precise, but it will
1) introduce more complex design
2) make it much harder to debug
3) slightly harm the performance (due to repetitive synchronization)

That's the reason why I chose a simplified implementation: it gives us correct transformation results fast and easily, while providing a rough estimation of exit condition that is non-essential.

Reply:
Thanks for your replies, I see your point.